processing/src/main/java/io/druid/query/aggregation/SuppressedAggregatorFactory.java

```diff
@@ -0,0 +1,375 @@
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This AggregatorFactory is meant for wrapping delegate aggregators for optimization purposes.
+ *
+ * The wrapper suppresses the aggregate() method for the underlying delegate, while leaving
+ * the behavior of other calls unchanged.
+ *
+ * This wrapper is meant to be used when an optimization decides that an aggregator can be entirely skipped
+ * (e.g., a FilteredAggregatorFactory where the filter condition will never match).
+ */
+public class SuppressedAggregatorFactory extends AggregatorFactory
+{
+  private final AggregatorFactory delegate;
+
+  public SuppressedAggregatorFactory(
+      AggregatorFactory delegate
+  )
+  {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    return new SuppressedAggregator(delegate.factorize(metricFactory));
+  }
+
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    return new SuppressedBufferAggregator(delegate.factorizeBuffered(metricFactory));
+  }
+
+  @Override
+  public Comparator getComparator()
+  {
+    return delegate.getComparator();
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    return delegate.combine(lhs, rhs);
+  }
+
+  @Override
+  public AggregateCombiner makeAggregateCombiner()
+  {
+    return delegate.makeAggregateCombiner();
+  }
+
+  @Override
+  public AggregatorFactory getCombiningFactory()
+  {
+    return delegate.getCombiningFactory();
+  }
+
+  @Override
+  public AggregatorFactory getMergingFactory(AggregatorFactory other) throws AggregatorFactoryNotMergeableException
+  {
+    return delegate.getMergingFactory(other);
+  }
+
+  @Override
+  public List<AggregatorFactory> getRequiredColumns()
+  {
+    return delegate.getRequiredColumns();
+  }
+
+  @Override
+  public Object deserialize(Object object)
+  {
+    return delegate.deserialize(object);
+  }
+
+  @Override
+  public Object finalizeComputation(Object object)
+  {
+    return delegate.finalizeComputation(object);
+  }
+
+  @Override
+  public String getName()
+  {
+    return delegate.getName();
+  }
+
+  @Override
+  public List<String> requiredFields()
+  {
+    return delegate.requiredFields();
+  }
+
+  @Override
+  public String getTypeName()
+  {
+    return delegate.getTypeName();
+  }
+
+  @Override
+  public int getMaxIntermediateSize()
+  {
+    return delegate.getMaxIntermediateSize();
+  }
+
+  @Override
+  public AggregatorFactory optimizeForSegment(PerSegmentQueryOptimizationContext optimizationContext)
+  {
+    // we are already the result of an optimizeForSegment() call
+    return this;
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] delegateCacheKey = delegate.getCacheKey();
+    return ByteBuffer.allocate(1 + delegateCacheKey.length)
```

Comment:
Please use `CacheKeyBuilder`. `AggregatorFactory` is `Cacheable`, so that you can simply use `builder.appendCacheable()`.

Reply:
Changed to use CacheKeyBuilder